src/main/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableBufferTimed.java

```diff
@@ -0,0 +1,573 @@
+            if (enter()) {
+                QueueDrainHelper.drainMaxLoop(queue, downstream, false, null, this);
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            requested(n);
+        }
+
+        @Override
+        public void cancel() {
+            cancelled = true;
+            upstream.cancel();
+            DisposableHelper.dispose(timer);
+        }
+
+        @Override
+        public void run() {
+            U next;
+
+            try {
+                next = Objects.requireNonNull(bufferSupplier.get(), "The supplied buffer is null");
+            } catch (Throwable e) {
+                Exceptions.throwIfFatal(e);
+                cancel();
+                downstream.onError(e);
+                return;
+            }
+
+            U current;
+
+            synchronized (this) {
+                current = buffer;
+                if (current == null) {
+                    return;
+                }
+                buffer = next;
+            }
+
+            fastPathEmitMax(current, false, this);
+        }
+
+        @Override
+        public boolean accept(Subscriber<? super U> a, U v) {
+            downstream.onNext(v);
+            return true;
+        }
+
+        @Override
+        public void dispose() {
+            cancel();
+        }
+
+        @Override
+        public boolean isDisposed() {
+            return timer.get() == DisposableHelper.DISPOSED;
+        }
+    }
+
+    static final class BufferSkipBoundedSubscriber<T, U extends Collection<? super T>>
+    extends QueueDrainSubscriber<T, U, U> implements Subscription, Runnable {
+        final Supplier<U> bufferSupplier;
+        final long timespan;
+        final long timeskip;
+        final TimeUnit unit;
+        final Worker w;
+        final List<U> buffers;
+
+        Subscription upstream;
+
+        BufferSkipBoundedSubscriber(Subscriber<? super U> actual,
+                Supplier<U> bufferSupplier, long timespan,
+                long timeskip, TimeUnit unit, Worker w) {
+            super(actual, new MpscLinkedQueue<>());
+            this.bufferSupplier = bufferSupplier;
+            this.timespan = timespan;
+            this.timeskip = timeskip;
+            this.unit = unit;
+            this.w = w;
+            this.buffers = new LinkedList<>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (!SubscriptionHelper.validate(this.upstream, s)) {
+                return;
+            }
+            this.upstream = s;
+
+            final U b; // NOPMD
+
+            try {
+                b = Objects.requireNonNull(bufferSupplier.get(), "The supplied buffer is null");
+            } catch (Throwable e) {
+                Exceptions.throwIfFatal(e);
+                w.dispose();
+                s.cancel();
+                EmptySubscription.error(e, downstream);
+                return;
+            }
+
+            buffers.add(b);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 This line may not be thread-safe: **269**. At this line the following field is not protected by synchronization: `buffers`. Your code typically protects this field with synchronization on object: `this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `this`. For example, like at lines: **375**, **329**, **283**, **302**, and **353**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **283**, **302**, **329**, **353**, and **375**.